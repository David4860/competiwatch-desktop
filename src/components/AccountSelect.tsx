import React, { useState } from "react";
import Account from "../models/Account";

interface Props {
  activeAccount: Account | null;
  onChange: (val: string) => void;
  onPageChange: (activePage: string, val1?: any, val2?: any) => void;
  accounts: Array<Account>;
}

const AccountSelect = ({
  activeAccount,
  onChange,
  onPageChange,
  accounts
}: Props) => {
  const [isOpen, setIsOpen] = useState(false);

  if (accounts.length < 1) {
    return null;
  }

  const containerClass = () => {
    const classes = ["select-menu", "d-inline-block"];
    if (isOpen) {
      classes.push("active");
    }
    return classes.join(" ");
  };

  const toggleButtonClass = () => {
    const classes = ["btn", "select-menu-button"];
    if (isOpen) {
      classes.push("selected");
    }
    return classes.join(" ");
  };

  const accountButtonClass = (accountID: string) => {
    const classes = ["select-menu-item", "text-left", "width-full", "btn-link"];
    if (activeAccount && activeAccount._id === accountID) {
      classes.push("selected");
    }
    return classes.join(" ");
  };

  const onAccountClick = (
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    const button = event.currentTarget;

    button.blur();
    onChange(button.value);
    setIsOpen(false);
  };

  const manageAccounts = () => {
    setIsOpen(false);
    onPageChange("accounts");
  };

  return (
    <div className={containerClass()}>
      <button
        className={toggleButtonClass()}
        type="button"
        onClick={() => setIsOpen(!isOpen)}
        aria-haspopup="true"
        aria-expanded="false"
      >
        {activeAccount ? activeAccount.battletag : "Select an account"}
      </button>
      <div className="select-menu-modal-holder">
        <div className="select-menu-modal">
          <div className="select-menu-list">
            {accounts.map(account => (
              <button
                className={accountButtonClass(account._id)}
                key={account._id}
                type="button"
                value={account._id}
                onClick={onAccountClick}
              >
                <span className="ion ion-ios-checkmark select-menu-item-icon" />
                <span className="select-menu-item-text">
                  {account.battletag}
                </span>
              </button>
            ))}
            <button
              className="select-menu-item text-bold text-left width-full btn-link"
              type="button"
              onClick={manageAccounts}
            >
              <span className="ion ion-ios-checkmark select-menu-item-icon" />
              <span className="select-menu-item-text">Manage accounts</span>
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

export default AccountSelect;
