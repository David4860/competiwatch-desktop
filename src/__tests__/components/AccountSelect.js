import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

const accounts = [
  { _id: "456", battletag: "CheshireCat" },
  { _id: "123", battletag: "MarchHare" }
];

global.window.ipcRenderer = {
  once: (_, handler) => {
    handler(null, null, { _id: "123", battletag: "MarchHare" });
  },
  send: () => null
};
jest.mock("is-electron", () => ({
  __esModule: true,
  default: () => true
}));
import AccountSelect from "../../components/AccountSelect";

it("matches snapshot", () => {
  const tree = renderer
    .create(<AccountSelect accounts={accounts} activeAccountID="123" />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders", async () => {
  const wrapper = shallow(
    <AccountSelect accounts={accounts} activeAccountID="123" />
  );
  const button = wrapper.find(".select-menu-button");
  expect(button.text()).toBe("Select an account");
});
