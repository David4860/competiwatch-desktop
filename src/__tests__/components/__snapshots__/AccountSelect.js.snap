// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="select-menu d-inline-block"
>
  <button
    aria-expanded="false"
    aria-haspopup="true"
    className="btn select-menu-button"
    onClick={[Function]}
    type="button"
  >
    Select an account
  </button>
  <div
    className="select-menu-modal-holder"
  >
    <div
      className="select-menu-modal"
    >
      <div
        className="select-menu-list"
      >
        <button
          className="select-menu-item text-left width-full btn-link"
          onClick={[Function]}
          type="button"
          value="456"
        >
          <span
            className="ion ion-ios-checkmark select-menu-item-icon"
          />
          <span
            className="select-menu-item-text"
          >
            CheshireCat
          </span>
        </button>
        <button
          className="select-menu-item text-left width-full btn-link"
          onClick={[Function]}
          type="button"
          value="123"
        >
          <span
            className="ion ion-ios-checkmark select-menu-item-icon"
          />
          <span
            className="select-menu-item-text"
          >
            MarchHare
          </span>
        </button>
        <button
          className="select-menu-item text-bold text-left width-full btn-link"
          onClick={[Function]}
          type="button"
        >
          <span
            className="ion ion-ios-checkmark select-menu-item-icon"
          />
          <span
            className="select-menu-item-text"
          >
            Manage accounts
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
